### ISSUE TITLE
Tools should not be permitted to create tens of thousands of outputs

### ISSUE THREAD

**Issue by natefoo** [galaxyproject/galaxy#13101]
**Describe the bug**
Tools which produce dynamic outputs can cause massive amounts of handler memory consumption if the tool produces tens of thousands of outputs or more. A frequent culprit on usegalaxy.org is the `rbc_splitfasta` tool. Typically I have to kill these jobs manually because they cause the handler to go into a memory crash loop until the job is manually set to a terminal state.

Tools should probably be updated to refuse to produce so many outputs, but we should also probably provide a way to configure a limit in Galaxy for tools that do not do it themselves. Even if we fixed the memory consumption issue, we're going to end up with even worse database bloat when it's almost certainly a misuse of the tool in the first place.

`rbc_splitfasta` is used as an example here but other tools that operate in the same manner can run in to the same problems, e.g. split file to collection, etc.

**Galaxy Version and/or server at which you observed the bug**
All recent versions.

**To Reproduce**
1. Install [rbc_splitfasta](https://toolshed.g2.bx.psu.edu/repository?repository_id=3538f67b895ead42)
2. Upload a fasta file with ~1M sequences
3. Run `rbc_splitfasta` with *Each sequence in its own dataset* selected for *Split Mode*
4. Enjoy the fireworks

**Expected behavior**
Tool quickly fails with a message that it cannot be used on the file. Or even better, the UI does not allow the input to be selected since Galaxy knows the number of sequences from the fasta metadata...

**Screenshots**
N/A

**Additional context**
N/A

**Comment by mvdbeek** [comment]
> Tool quickly fails with a message that it cannot be used on the file. Or even better, the UI does not allow the input to be selected since Galaxy knows the number of sequences from the fasta metadata...

I agree that would be nice, but there's no way (today) that Galaxy will know how many datasets are going to be created from the input, since that's all logic happening in the command section.

I think the only thing we can reasonably do without adding a new language for describing the expected number of outputs given some input metadata (... sounds both cool and challenging ...) is to fail at some cutoff of discovered files. And that's after the job has run, because we can't know this in advance.

Something to consider for allowing the transparent use of such tools is to create some kind of transparent collection that is just a single entry in the database, but that is indexed and can be unpacked like a regular collection ... or a directory datatype as we use in the CWL branch. Which would still possibly result in a huge number of downstream map over jobs, but we could try being smart about this and batch them together if individual inputs are really small.

Long story short, I can fail these jobs for now with a message explaning why they failed.

**Comment by mvdbeek** [comment]
Of course that's a global limit that could later come back to bite us ... it might not be worth it considering the split tools are usually not very resource intensive

**Comment by mvdbeek** [comment]
Now that we have https://github.com/galaxyproject/galaxy/pull/13170 i'll drop the bug label and the milestone. Getting something like a dynamic collection would be really cool, but also tricky

**Comment by hexylena** [comment]
Just to add on to this, it would be great if we have a way to prevent creating those files in the first place - some storage backends really suffer under 100k+ files in a single directory.

**Comment by mvdbeek** [comment]
> we have a way to prevent creating those files in the first place

They're not precreated in the object store, only in the job working directory. Is that also problematic ?

**Comment by hexylena** [comment]
Depends on the backing file store where the working directory lives. I know it's been a problem for @bgruening in the past when the working directory was EMC Isilon backed rather than on scratch space.

**Comment by mvdbeek** [comment]
Resolved by #13170.
